main.hs

```diff
@@ -3,171 +3,445 @@
+            pageData <-
+                trace (" * Reading data from Wiki page ID: " ++ pageId)
+                getData pageId
+            case pageData of
+                Nothing -> liftIO $ putStrLn "!!! Warning: failed to fetch doc, not updating the listings..."
+                Just pageData' -> do
+                    table <- parseTable pageData'
+                    forM_ languages $ \lang ->
+                        trace (" *     Language: " ++ show lang)
+                        renderTable rootDir lang pc table
+
+
+-- =============================================================================
 -- * Types
+-- =============================================================================
+
 
+-- TODO: Why is it using the three types, ReaderT Config IO
+-- | Short hand for the combination of the three types.
 type M = ReaderT Config IO
-type Lang = Text -- ^ en, se, fi, ...
-data PageConf = PageConf
-              { pageId    :: String
-              , pageUrl   :: Map Lang Text
-              , pageTitle :: Map Lang Text
-              } deriving Generic
-data Config   = Config
-              { fetchUrl, cacheDir, oodiNameFile
-              , rootDir                           :: FilePath
-              , weboodiUrl                        :: Text
-              , pages                             :: [PageConf]
-              , colCode, colLang, colCourseName
-              , colRepeats, colPeriod, colWebsite
-              , colLukukausi, classCur            :: Text
-              , categories                        :: [[Text]]
-              , i18n                              :: I18N
-              , languages                         :: [Lang]
-              } deriving Generic
 
+
+{- | The 'Lang' type is used as key for looking up the translations from the
+    internationalization ('I18N') database found in the /config.yaml/ file.
+
+    It is used with acronyms of the languages: @fi@, @se@, @en@, ...
+-}
+type Lang = Text
+
+
+{- | Properties for generating individual web page bodies.
+
+    The 'PageConf' data type is used as a data holder for page information
+    of the generated web pages. Where they are stored locally, what the page
+    title is etc.
+-}
+data PageConf = PageConf {
+              -- Department web page properties
+                pageId          :: String         -- ^ Wiki Table ID
+              , pageUrl         :: Map Lang Text  -- ^ Local path
+              , pageTitle       :: Map Lang Text  -- ^ Local title
+              } deriving Generic
 instance Yaml.FromJSON PageConf
+
+
+{- | Overall properties used by the module to generate HTML code from the
+    source 'Table' found in the wiki pages.
+
+    The 'Config' data type is used to read the configuration properties from
+    the /config.yaml/ file. It holds references to all of the different
+    property fields and are accessed by their name in the /config.yaml/ file.
+    The fields are filled with a call to the @'ask'@ function.
```

Comment:
En ymmärrä viimeistä lausetta "The fields are filled with a call to the `ask` function." Mitkä kentät tarkalleen? datatyypin kentät kun täytetään ihan normaalisti Config-konstruktorilla